### ISSUE TITLE
Extend User Info Based on Identity Provider

### ISSUE THREAD

**Issue by b-f-chan** [overture-stack/ego#507]
We will add a global config on an Ego  instance that controls whether `Email` has a unique constraint (True/False) [#518]

For the sake of this tickets development, we will be assuming that the email constraint has been set to `True` as that is how Ego is currently running, and that is also required for Argo development. 

We need to: 
1. Add a `unique-id` to the UserInfo Table, as well as an `identityProvider` field. these should be 1:1 right now (a user has 1 id provider; no account merging
2. This  `unique-id`+ `identityProvider` will form a composite key to identify people.  Once this is in place, we can remove the unique constraint on the the email field. 
3. right now both Email and name are non-nullable.  Once the unique constraint is removed, allow these to be nullable. 
4. Add logic such that  `unique-id`+ `identityProvider` are never empty 
--- when a user is created populate them 
--- if a user exists without these, when they login ensure that they are populated
5. Update the SOO filter to query for ID provider, and branch logic based on chosen ID provider, specifically:

Expand the logic for userService.getUserByToken to take the following steps:
Check for an existing user with the incoming idToken.providerType + idToken.providerId.
If #1 fails, check by the idToken.email as providerId + the idToken.providerType. If a user is found this way, update their providerType and providerId with the values in the idToken*
If #2 fails, a new user will be created with provider info populated
*In the case of ORCID, we’re not guaranteed that an email will be present in the idToken. If a user doesn’t have their email visible in the idToken we would need to create a new user as we can’t be certain of finding a matching user account

**Comment by akachru-github** [comment]
Will likely be a migration involved with existing users already in Ego. Discuss in planning.

**Comment by b-f-chan** [comment]
The `name` field needs to be removed from the User Entity as well as part of this ticket scope, since it no longer applies.  It needs to be removed from the model and some logic around that field may also need to be removed.  See the _Ego ID Provider Corrections & Consequences_ doc for details.  This increase the complexity of the task, this change was discovered naturally as part of actually doing the ticket and finding out there is more to the task than originally understood.

**Comment by anncatton** [comment]
We'll also be removing the user `POST` endpoint as the provider info cannot be known without a user's consent.

**Comment by b-f-chan** [comment]
Before proceeding with the expanded ticket in next sprint, we'll touch base at planning as a group so we're all on the same page with the changes

**Comment by b-f-chan** [comment]
Scope of this ticket expanded as implementation proceeded and more discovery was made of extra tasks that were needed to support this.  These tasks are needed to properly implement a solution where EGO can differentiate users from different ID providers.

Some tasks were mandatory to complete this ticket, others identified as future work that can be taken up in coming sprints (but still necessary).  To break this ticket up and make it more manageable, we will track some of these as separate tasks:

1. Remove createUser endpoint from EGO.  This is required because in the new implementation users must be created via the OIDC flow and with users' consent.  The current createUser endpoint allows a bypass for this and needs to be removed or suppressed.  See ticket <>

2. Remove Name field from User table in EGO.  This is required because the Name column is no longer useful as a unique identifier in the new implementation.  The composite of `providerId` + `providerType` are now used to uniquely identify an account.  See ticket <>.

3. Add a logic table to check, upon user sign-in, if a user exists based on the incoming Token's `providerId` + `providerType`.  If this fails, check for an existing user instead with the e-mail as the `providerId` and if found, update both fields with the Token's values.  If this still fails, then create a new user with the Token's `providerId` + `providerType`.   This is required for migration when we put the new implementation into production.  On initial release, we only know existing users have a unique e-mail and default provider must be Google.  So when they sign in again, we need to make sure their provider details are updated with the Token's information.  See ticket <>.

4. Based on the logic table in #3, update the userUpdate endpoint to make use of this logic.  This is required to complete the behaviour in #3 above.  See ticket <>.

**Comment by b-f-chan** [comment]
Testing for all these requirement was delegated out to separate tickets, this can be closed now that all related tickets are successfully tested and in staging
